feat: add `eslint/rules/eol-open-bracket-spacing`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add `ESLint` rule for `eol-open-bracket-spacing`
-   add `fixer` function to auto-fix the rule.

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add eslint/rules/eol-open-bracket-spacing

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on etc/eslint/rules/stdlib.js:
```diff
@@ -200,6 +200,52 @@ rules[ 'stdlib/doctest-quote-props' ] = 'error';
 rules[ 'stdlib/empty-line-before-comment' ] = 'error';
 
+/**
+* No spaces allowed between an opening parenthesis or bracket and a nested object or array expression.
```
```suggestion
* Disallow spaces between an opening parenthesis or bracket and a nested object or array expression at the end of a line.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/README.md:
```diff
@@ -0,0 +1,165 @@
+var Linter = require( 'eslint' ).Linter;
+var rule = require( '@stdlib/_tools/eslint/rules/eol-open-bracket-spacing' );
+
+var linter = new Linter();
+var result;
+var code;
+
+code = [
+    'function test() {',
+    '  var log = require( \'@stdlib/console/log\' );',
+    '  log( {',
+    '    "key": "value"',
+    '  });',
+    '  var arr = [ {',
+    '    "nested": true',
+    '  }];',
+    '  log( arr );',
+    '}'
+].join( '\n' );
+
+linter.defineRule( 'eol-open-bracket-spacing', rule );
+
+result = linter.verify( code, {
```
```suggestion
var linter = new Linter();

var code = [
    'function test() {',
    '  var log = require( \'@stdlib/console/log\' );',
    '  log( {',
    '    "key": "value"',
    '  });',
    '  var arr = [ {',
    '    "nested": true',
    '  }];',
    '  log( arr );',
    '}'
].join( '\n' );

linter.defineRule( 'eol-open-bracket-spacing', rule );

var result = linter.verify( code, {
```

The previous style should be considered "legacy".

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/README.md:
```diff
@@ -0,0 +1,165 @@
+
+# eol-open-bracket-spacing
+
+> [ESLint rule][eslint-rules] to enforce that no spaces are present between an opening parenthesis or bracket and a nested object or array expression.
```
```suggestion
> [ESLint rule][eslint-rules] to enforce that no spaces are present between an opening parenthesis or bracket and a nested object or array expression at the end of a line.
```

Change applies here and elsewhere. Otherwise, the description reads as if `[ { 'a': 1 } ]` (i.e., on the same line) also applies.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/README.md:
```diff
@@ -0,0 +1,165 @@
+        {
+            'ruleId': 'eol-open-bracket-spacing',
+            'severity': 2,
+            'message': 'No spaces allowed between an opening parenthesis or bracket and a nested object or array expression',
```
```suggestion
            'message': 'No spaces allowed between an opening parenthesis or bracket and a nested object or array expression at the end of a line',
```

Applies here and below.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/examples/index.js:
```diff
@@ -0,0 +1,68 @@
+        {
+            'ruleId': 'eol-open-bracket-spacing',
+            'severity': 2,
+            'message': 'No spaces allowed between an opening parenthesis or bracket and a nested object or array expression',
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/lib/main.js:
```diff
@@ -0,0 +1,158 @@
+* @returns {Object} validators
+*/
+function main( context ) {
+	var source;
+
+	source = context.getSourceCode();
```
```suggestion
	var source = context.getSourceCode();
```

For a trivial case like this (single variable, no other logic), it is fine to declare and assign on the same line.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/lib/main.js:
```diff
@@ -0,0 +1,158 @@
+	function report( node, prevToken, tokenAfter ) {
+		context.report({
+			'node': node,
+			'message': 'No spaces allowed between an opening parenthesis or bracket and a nested object or array expression',
```
```suggestion
			'message': 'No spaces allowed between an opening parenthesis or bracket and a nested object or array expression at the end of a line',
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/lib/main.js:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	/**
+	* Checks whether there are spaces present between an opening parenthesis or bracket and a nested object or array expression.
```
```suggestion
	* Checks whether there are spaces present between an opening parenthesis or bracket and a nested object or array expression at the end of a line.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/lib/main.js:
```diff
@@ -0,0 +1,158 @@
+	'meta': {
+		'type': 'layout',
+		'docs': {
+			'description': 'enforce no spaces between an opening parenthesis or bracket and a nested object or array expression'
```
```suggestion
			'description': 'disallow spaces between an opening parenthesis or bracket and a nested object or array expression at the end of a line'
```

kgryte reviewed (changes requested):
Thanks for working on this, @headlessNode. This is looking great! Just a few small nits and this should be ready to merge!

headlessNode committed:
refactor: apply code review suggestions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/README.md:
```diff
@@ -0,0 +1,163 @@
+
+#### rule
+
+[ESLint rule][eslint-rules] to enforce that no spaces are present between an opening parenthesis or bracket and a nested object or array expression.
```
```suggestion
[ESLint rule][eslint-rules] to enforce that no spaces are present between an opening parenthesis or bracket and a nested object or array expression at the end of a line.
```

headlessNode commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/package.json:
```diff
@@ -0,0 +1,63 @@
+{
+  "name": "@stdlib/_tools/eslint/rules/eol-open-bracket-spacing",
+  "version": "0.0.0",
+  "description": "ESLint rule to enforce that no spaces are present between an opening parenthesis or bracket and a nested object or array expression.",
```
```suggestion
  "description": "ESLint rule to enforce that no spaces are present between an opening parenthesis or bracket and a nested object or array expression at the end of a line.",
```

headlessNode commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/test/test.js:
```diff
@@ -0,0 +1,86 @@
+	t.end();
+});
+
+tape( 'the function positively validates code where no spaces are present between an opening parenthesis or bracket and a nested object or array expression', function test( t ) {
```
```suggestion
tape( 'the function positively validates code where no spaces are present between an opening parenthesis or bracket and a nested object or array expression at the end of a line', function test( t ) {
```

headlessNode commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/test/test.js:
```diff
@@ -0,0 +1,86 @@
+	t.end();
+});
+
+tape( 'the function negatively validates code where spaces are present between an opening parenthesis or bracket and a nested object or array expression', function test( t ) {
```
```suggestion
tape( 'the function negatively validates code where spaces are present between an opening parenthesis or bracket and a nested object or array expression at the end of a line', function test( t ) {
```

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/_tools/eslint/rules/eol-open-bracket-spacing/examples/index.js:
```diff
@@ -0,0 +1,68 @@
+var rule = require( './../lib' );
+
+var linter = new Linter();
+var result;
+var code;
+
+code = [
+	'function test() {',
+	'  var log = require( \'@stdlib/console/log\' );',
+	'  log( {',
+	'    "key": "value"',
+	'  });',
+	'  var arr = [ {',
+	'    "nested": true',
+	'  }];',
+	'  log( arr );',
+	'}'
+].join( '\n' );
+
+linter.defineRule( 'eol-open-bracket-spacing', rule );
+
+result = linter.verify( code, {
+	'rules': {
+		'eol-open-bracket-spacing': 'error'
+	}
+});
```
```suggestion

var code = [
	'function test() {',
	'  var log = require( \'@stdlib/console/log\' );',
	'  log( {',
	'    "key": "value"',
	'  });',
	'  var arr = [ {',
	'    "nested": true',
	'  }];',
	'  log( arr );',
	'}'
].join( '\n' );

linter.defineRule( 'eol-open-bracket-spacing', rule );

var result = linter.verify( code, {
	'rules': {
		'eol-open-bracket-spacing': 'error'
	}
});
```

headlessNode committed:
refactor: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented:
@kgryte Changes applied. Please review.

kgryte reviewed (approved):
LGTM. Thanks, @headlessNode!
